### ISSUE TITLE
Access to request properties other than slots

### ISSUE THREAD

**Issue by tjaffri** [cameronhunter/alexa-annotations#22]
These annotations are fantastic if you are dealing with one of two cases:

1. A simple request: launch, session ended, etc... with no parameters.
2. An intent with any number of slots, e.g. something like this:

```json
"request": {
        "type": "IntentRequest",
        "requestId": "amzn1.echo-api.request.47ed6652-a1af-4c36-ade6-59090b1ac0a1",
        "timestamp": "2016-10-06T04:19:23Z",
        "locale": "en-US",
        "intent": {
            "name": "PlayVerseByNumbersIntent",
            "slots": {
                "ChapterNumber": {
                    "name": "ChapterNumber",
                    "value": "1"
                },
                "VerseNumber": {
                    "name": "VerseNumber",
                    "value": "3"
                }
            }
        }
    }
```

In the example above, you could have something like the following to parse out the slots in your skill code:

```js
 @Intent('PlayVerseByNumbersIntent')
  async playVerseByNumbers({ ChapterNumber, VerseNumber }) {
      // do something with the given ChapterNumber and VerseNumber
  }
```

However, what if you need access to one of the other properties of the request? e.g. the requestId, locale or token? The last one, token, is especially important for dealing with audio requests... e.g. when you play something and your audio playback has almost finished, Alexa sends you this request:

```js
  "request": {
        "type": "AudioPlayer.PlaybackNearlyFinished",
        "requestId": "amzn1.echo-api.request.0023cee1-52e3-465c-97cd-e8bb36c624f6",
        "timestamp": "2016-10-06T04:19:30Z",
        "locale": "en-US",
        "token": "{\"ChapterNumber\":\"1\",\"VerseNumber\":\"3\"}",
        "offsetInMilliseconds": 66
  }
```

The token field is passed-through from the original play directive that was sent to play the item that has almost finished playing. In this case, I would like to parse the token and queue up the NEXT item (token is free-form, and can be anything... I just chose to make it a JSON containing context about the item).

Is there an easy way to access other request properties in cases like AudioPlayer.PlaybackNearlyFinished? I was thinking of adding a new annotation type ``@RawRequest`` so that somebody can do this if they want for advanced cases:

```js
 @RawRequest('AudioPlayer.PlaybackNearlyFinished')
  async audioPlayerPlaybackNearlyFinished(request) {
      // do something with the raw request, e.g. access request.requestId or request.token
  }
```

What do you think? I can send a PR if this looks reasonable to you. Some alternative designs would be:

1. Somehow pass the request object to EVERY request (perhaps as the first parameter followed by any slots etc?)
2. Rather than @RawRequest create something special for audio requests, e.g. @AudioPlayer that is bespoke for audio player requests... one of the initial features would be to pass the raw request info.

**Comment by cameronhunter** [comment]
The `request` is available as the second parameter to each annotated function:

```javascript
@Intent('PlayVerseByNumbersIntent')
async playVerseByNumbers(slots, request) {
  const { ChapterNumber, VerseNumber } = slots;
  // ... do something with slots or request
}
```

Oddly, the request is *always* the second parameter, even when slot data doesn't make sense:

```javascript
@Launch
async onLaunch(_, request) {
  // ... do something with request
}
```

It would be good to provide only the request to `@Launch` and `@SessionEnded`.

**Comment by tjaffri** [comment]
Awesome! I'll give it a shot... will keep this issue open for now in case I see something else in this space.

**Comment by cameronhunter** [comment]
I've just updated the package so that `request` is passed as the first parameter to `@Launch` and `@SessionEnded`. It continues to be the second parameter for the `@Intent` annotation:

```javascript
@Skill
class Example {
  @Launch
  onLaunch: (request) => {},

  @Intent('AMAZON.HelpIntent')
  onHelp: (slots, request) => {},

  @SessionEnded
  onSessionEnded: (request) => {}
}
```

This is published to npm as `v7.0.0`

**Comment by cameronhunter** [comment]
With regards to audio intents, it might be interesting to add another annotation for them. I would be open to exposing the `annotation` factory to make plugins like this possible.
